R-notebooks/covidhub_evaluation.Rmd

```diff
@@ -49,57 +49,97 @@ forecasters = c("CMU-TimeSeries",
 for (i in 1:length(forecasters)) {
   forecast_dates[[i]] = tryCatch({
-    get_forecast_dates(forecasters[i])
+    as.Date(get_forecast_dates(forecasters[i]))
   },
   error = function(e) cat(sprintf("%i. %s\n", i, e$message))
   )
 }
 ```
 
+Load data from previous run so we don't have to reingest / process it. This
+data could end up out of date if a forecast is retrospectively updated, but in
+that case it's no longer a true prediction. We can always restart from scratch
+by deleting covidhub_evaluation.rda.
+```{r}
+load(file = "covidhub_evaluation.rda")
+if (exists("predictions_cards")){
+  seen_dates = list()
+  for (i in 1:length(predictions_cards)){
+    forecaster_predictions_cards = predictions_cards[[i]]
+    pred_dates = lapply(forecaster_predictions_cards, 
+                        FUN = function(x) attributes(x)$forecast_date)
+    seen_dates[[i]] = unique(do.call("c", pred_dates))
+    name_of_forecaster = attributes(
+                          forecaster_predictions_cards[[1]])$name_of_forecaster
+    names(seen_dates)[[i]] = name_of_forecaster
+  }
+} else {
+  for (i in 1:length(forecasters)){
+    seen_dates[[i]] = list()
+  }
+  names(seen_dates) = forecasters
+}
+```
+
 Now figure out "comparable" forecast dates: making a forecast on a Sunday or a 
 Monday of the same epiweek should be comparable. **TODO**: we should switch over 
 to using the Zoltar API, and soon it should have an "as of" parameter, so then 
 we shouldn't need to do this. 
 
-```{r, eval = FALSE}
-forecast_dates_comparable = vector("list", length = length(forecasters)) 
+```{r}
+
 forecast_dates_cmu = forecast_dates[[1]]
+new_dates = list()
 for (i in 1:length(forecasters)) {
   given_dates = forecast_dates[[i]]
-  for (j in 1:length(forecast_dates_cmu)) {
-    # Find the last forecast date before the current CMU one
-    given_date = given_dates[max(which(given_dates <= forecast_dates_cmu[j]))]
-    
-    # If the dates match exactly, or the given date falls on a Sunday and the
-    # CMU date falls on a Monday of the same epiweek, the call it comparable
-    if (!is.na(given_date) &&
-        (given_date == forecast_dates_cmu[j] ||
-        (as.Date(forecast_dates_cmu[j]) - as.Date(given_date) == 1 &&
-         wday(forecast_dates_cmu[j]) == 2 && wday(given_date) == 1))) {
-      forecast_dates_comparable[[i]][j] = given_date
-    }
-    
-    # Otherwise call it NA
-    else forecast_dates_comparable[[i]][j] = NA
+  # If the dates match exactly, or the given date falls on a Sunday and the
+  # CMU date falls on a Monday of the same epiweek, then call it comparable...
+  dates_comp = given_dates[(given_dates %in% forecast_dates_cmu | 
+                            ((given_dates + 1) %in% forecast_dates_cmu) &
+                            wday(given_dates) == 1)]
+  
+  # ...but if there is an exact match on dates, ignore predictions made on the
+  # previous day
+  extra_dates = dates_comp[(dates_comp + 1) %in% dates_comp]
```

Comment:
Took me a while to parse through this.  This variable stores all dates except for the last one in each continuous blocks?  I would like to see a more descriptive name, like `previous_{consecutive_}dates_with_forecasts`.

Reply:
>This variable stores all dates except for the last one in each continuous blocks

Yes, but the reason is a little bit more subtle.

CMU makes forecasts once a week. We consider a forecast date to be comparable to a CMU forecast if it was made on the same day or the previous day. So, if an organization makes weekly forecasts on a Sunday or Monday, we want to include it.

But, some organizations make forecasts daily and we don't want to include both the Sunday and Monday versions in that case, just the Monday version.

So, it is the last day in the chain, but it is really just excluding the Sunday version if the Monday version is also there.